src/core/batch.rkt

```diff
@@ -146,40 +159,56 @@
   (define roots (vector-map (curry vector-ref mapping) (batch-roots b)))
   (mutable-batch->batch out roots))
 
-; The function removes any zombie nodes from batch with respect to the roots
-; Time complexity: O(|R| + |N|), where |R| - number of roots, |N| - length of nodes
-; Space complexity: O(|N| + |N*| + |R|), where |N*| is a length of nodes without zombie nodes
-; The flag keep-vars is used in compiler.rkt when vars should be preserved no matter what
-(define (batch-remove-zombie input-batch [roots (batch-roots input-batch)] #:keep-vars [keep-vars #f])
-  (define nodes (batch-nodes input-batch))
-  (define nodes-length (batch-length input-batch))
+;; Function returns indices of alive nodes within a batch for given roots,
+;;   where alive node is a child of a root + meets a condition - (condition node)
+(define (batch-alive-nodes batch
+                           [roots (batch-roots batch)]
+                           #:keep-vars-alive [keep-vars-alive #f]
+                           #:condition [condition (const #t)])
+  (define nodes (batch-nodes batch))
+  (define nodes-length (batch-length batch))
+  (define alive-mask (make-vector nodes-length #f))
+  (for ([root (in-vector roots)])
+    (vector-set! alive-mask root #t))
+  (for/vector ([i (in-range (- nodes-length 1) -1 -1)]
+               [node (in-vector nodes (- nodes-length 1) -1 -1)]
+               [alv (in-vector alive-mask (- nodes-length 1) -1 -1)]
+               #:when (or (and alv (condition node)) (and keep-vars-alive (symbol? node))))
+    (expr-recurse node (λ (n) (vector-set! alive-mask n #t)))
+    i))
```

Comment:
Because you create the vector in reverse order you end up needing to do `vector-sort!`. Just add one extra loop at the end going through the alive-mask in order, and you won't need to sort the results.